deepchem/models/torch_models/gat.py

```diff
@@ -1,224 +1,338 @@
-  one of variants of Graph Convolutional Networks. The main difference between basic GCN models
-  is how to update node representations. The GAT uses multi head attention mechanisms which
-  outbroke in NLP like Transformer when updating node representations. The most important advantage
-  of this approach is that we can get the interpretability like how the model predict the value
-  or which part of the graph structure is important from attention-weight. Please confirm
-  the detail algorithms from [1]_.
+  * Update node representations in graphs with a variant of GAT
+  * For each graph, compute its representation by 1) a weighted sum of the node
+    representations in the graph, where the weights are computed by applying a
+    gating function to the node representations 2) a max pooling of the node
+    representations 3) concatenating the output of 1) and 2)
+  * Perform the final prediction using an MLP
 
   Examples
   --------
+
   >>> import deepchem as dc
-  >>> from torch_geometric.data import Batch
+  >>> import dgl
+  >>> from deepchem.models import GAT
   >>> smiles = ["C1CCC1", "C1=CC=CN=C1"]
   >>> featurizer = dc.feat.MolGraphConvFeaturizer()
   >>> graphs = featurizer.featurize(smiles)
   >>> print(type(graphs[0]))
   <class 'deepchem.feat.graph_data.GraphData'>
-  >>> pyg_graphs = [graph.to_pyg_graph() for graph in graphs]
-  >>> print(type(pyg_graphs[0]))
-  <class 'torch_geometric.data.data.Data'>
-  >>> model = dc.models.GAT(mode='classification', n_tasks=10, n_classes=2)
-  >>> preds, logits = model(Batch.from_data_list(pyg_graphs))
+  >>> dgl_graphs = [graphs[i].to_dgl_graph() for i in range(len(graphs))]
+  >>> # Batch two graphs into a graph of two connected components
+  >>> batch_dgl_graph = dgl.batch(dgl_graphs)
+  >>> model = GAT(n_tasks=1, mode='regression')
+  >>> preds = model(batch_dgl_graph)
   >>> print(type(preds))
   <class 'torch.Tensor'>
-  >>> preds.shape == (2, 10, 2)
+  >>> preds.shape == (2, 1)
   True
 
   References
   ----------
-  .. [1] Veličković, Petar, et al. "Graph attention networks." arXiv preprint
-     arXiv:1710.10903 (2017).
+  .. [1] Petar Veličković, Guillem Cucurull, Arantxa Casanova, Adriana Romero, Pietro Liò,
+         and Yoshua Bengio. "Graph Attention Networks." ICLR 2018.
 
   Notes
   -----
-  This class requires PyTorch Geometric to be installed.
+  This class requires DGL (https://github.com/dmlc/dgl) and DGL-LifeSci
+  (https://github.com/awslabs/dgl-lifesci) to be installed.
   """
 
-  def __init__(
-      self,
-      in_node_dim: int = 30,
-      hidden_node_dim: int = 32,
-      heads: int = 1,
-      dropout: float = 0.0,
-      num_conv: int = 2,
-      predictor_hidden_feats: int = 64,
-      n_tasks: int = 1,
-      mode: str = 'classification',
-      n_classes: int = 2,
-  ):
+  def __init__(self,
+               n_tasks: int,
+               graph_attention_layers: list = None,
```

Comment:
If you have already set None as the initial value of the arguments in GCNModel, it will be better to modify them.

Reply:
Many people consider this as a bad practice as it can lead to strange bugs. See https://medium.com/better-programming/mutable-default-arguments-in-python-643ae2583e00. This is why I used `None` for the default value and added an explanation in the doc.

Reply:
OK! I understood.